torch/optim/_functional.py

```diff
@@ -92,21 +92,35 @@ def adam(params: List[Tensor],
         if weight_decay != 0:
             grad = grad.add(param, alpha=weight_decay)
 
+        is_complex = torch.is_complex(param)
+        if is_complex:
+            grad = torch.view_as_real(grad)
+            exp_avg = torch.view_as_real(exp_avg)
+            exp_avg_sq = torch.view_as_real(exp_avg_sq)
+            param = torch.view_as_real(param)
+
         # Decay the first and second moment running average coefficient
         exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
```

Comment:
To update the `exp_avg` in `state`, do I need to return `exp_avg` to the `state['exp_avg']` after `exp_avg = torch.view_as_real(exp_avg)` and `exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)`?

Reply:
No need because the `exp_avg` is the same Tensor as the one that is in `state`. And since you modify `exp_avg` inplace below, then you are by side effect modifying the Tensor that is in `state` as well.

Reply:
OK, thanks, interesting side effects lol.